### PR TITLE
postgresql_user: add groups parameter

### PR DESCRIPTION
**Pull request by Andersson007**
##### SUMMARY
postgresql_user: add groups parameter to be able to grant roles to the user

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
postgresql_user

##### ADDITIONAL INFORMATION
```
- name: Create user test and grant group user_ro and user_rw to it
  postgresql_user:
    name: test
    groups:
    - user_ro
    - user_rw
```

### CODE REVIEW

**Comment by Andersson007** [comment]
ready_for_review

**Comment by Andersson007** [comment]
ready_for_review

File: lib/ansible/modules/database/postgresql/postgresql_user.py [hunk 1]
```diff
@@ -750,6 +763,91 @@ def get_valid_flags_by_version(cursor):
...
+        self.changed = False
+        self.__check_roles_exist()
+
+    def grant(self):
+        for group in self.groups:
+            self.granted[group] = []
+
+            for role in self.target_roles:
+                # If role is in a group now, pass:
+                if self.__check_membership(group, role):
```
Comment (felixfontein) [inline lib/ansible/modules/database/postgresql/postgresql_user.py:784]: If `groups` and `target_roles` are longer lists, this will be pretty slow: it does `len(self.groups) * len(self.target_roles)` queries. Wouldn't it be better to make one query which selects the group for every role in `self.target_roles` (outside both loops), and then check for each role whether the group in the result for it equals `group`?

Comment (felixfontein) [reply lib/ansible/modules/database/postgresql/postgresql_user.py:784]: (Or maybe I misunderstood something, but this can probably be done more efficiently by doing at most one query per outer loop iteration.)

File: lib/ansible/modules/database/postgresql/postgresql_user.py [hunk 2]
```diff
@@ -750,6 +763,91 @@ def get_valid_flags_by_version(cursor):
     ]
 
 
+class PgMembership():
+    def __init__(self, module, cursor, target_roles, groups, fail_on_role=True):
```
Comment (felixfontein) [inline lib/ansible/modules/database/postgresql/postgresql_user.py:767]: Is there a reason the last optional option is called the same as a module option, but seems to be unrelated to it?

Comment (felixfontein) [reply lib/ansible/modules/database/postgresql/postgresql_user.py:767]: (Actually, this constructor argument is never set.)

**Review by felixfontein — COMMENTED** [review]
Changelog fragment :)

File: lib/ansible/modules/database/postgresql/postgresql_user.py (hunk 1 above)
Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_user.py:784]: the idea sounds good. However this class is almost the same as in postgresql_membership module. I'm going to move it to module_utils/postgresql.py and make it shared.
However I'd like to do it by particular PR.
Your idea is good, be sure i'll try to make this optimization soon for both modules at the same time (if i do it now it might lead me to confusion in the future and i'll have to rewrite it again)

File: lib/ansible/modules/database/postgresql/postgresql_user.py (hunk 2 above)
Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_user.py:767]: the same reason as in the comment above.

**Comment by Andersson007** [comment]
@felixfontein , I added the changelog fragment. It has CI for the case and it doesn't break existing tests. So, if there are not any other question, i believe it could be merged (i'd like to have more reviews, it's always cool, but the WG is a bit dead this summer.. unfortunately. Seems the guys don't want to hear anything about reviewing laying on the beaches, they hate me and ignore annoying notifications:)))

File: lib/ansible/modules/database/postgresql/postgresql_user.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/modules/database/postgresql/postgresql_user.py:784]: Sure, sounds good to me :)

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@Andersson007 thanks for implementing this new feature!

**Comment by Andersson007** [comment]
@felixfontein , thank you for reviewing and merging!
